_posts/2016-09-16-angular-2-animation-important-concepts

```diff
@@ -0,0 +1,236 @@
+@Component({
+  selector: 'my-fader',
+    template: `
+    <div *ngIf="visibility == 'shown'" >
+      <ng-content></ng-content>
+      Can you see me? 
+    </div>
+  `
+})
+export class MyComponent implements OnChanges {
+  visibility = 'shown';
+
+  @Input() isVisible : boolean = true;
+
+  ngOnChanges() {
+   this.visibility = this.isVisible ? 'shown' : 'hidden';
+  }
+}
+{% endraw %}
+{% endhighlight %}
+
+This component simply publishes an **@Input** `isVisible` property; which allows other external components to show/hide the text content... without any animations.
+
+<iframe style="width: 100%; height: 600px" src="http://embed.plnkr.co/chNcuU7IoIrN22DF" frameborder="0" allowfullscren="allowfullscren"></iframe>
+
+
+### Configure Component Animations
+
+We want the `my-fader` component to **fade-in** or **fade-out** its text content. And we want to *animate* those fades effects.
+
+> The essential take-away Animation concept is that **Angular 2 Animations** are triggered on component state changes. 
+Developers should consider state changes simply as value changes in a property of the component instance.
+
+To start animating, let's first add animation metadata to our component.
+
+Since our component has a `visibility` property whose state can change between `shown` and `hidden`, 
+let’s configure animations to trigger and animate during each value change.
+
+{% highlight js %}
+{% raw %}
+animations: [
+  trigger(’visibilityChanged', [
+  state('shown' , style({ opacity: 1 })), 
+  state('hidden', style({ opacity: 0 }))
+  ])
+]
+{% endraw %}
+{% endhighlight %}
+
+But what does the above mean... and why is this syntax used?
+
+The techno-speak above is saying that when the `visibilityChanged` property changes to the value == 'shown', 
+then the host element opacity changes to 1. And when the value changes == ‘hidden’, the 
+host element opacity should change to 0.
+
+And since we want to animate these changes instead of instantly hiding/showing the content, 
+we now need to configure a *transition*. 
+
+With Angular 2 this is also suprisingly easy:
+
+{% highlight js %}
+{% raw %}
+animations: [
+  trigger(’visibilityChanged', [
+    state('shown' , style({ opacity: 1 })), 
+    state('hidden', style({ opacity: 0 })),
+    transition('* => *', animate('.5s'))
+  ])
+]
+{% endraw %}
+{% endhighlight %}
+
+With the above `transition`, we added information to the animation configuration so each trigger value change will have a 500 msec transition. 
+
+So a **fade-in** (opacity 0 -> 1, with a duration of 500 msec) will occur when the value changes from `hidden` to `shown`. 
+And likewise the **fade-out** (opacity 1 -> 0, with a duration of 500 msec) will occur when the value changes from `shown` to `hidden`.
+
+I this this metadata notation is amazing:
+
+*  intuitive and easy to understand, 
+*  declarative, and 
+*  immediately associated with the component... 
+  *  the `animations` configuration is right above the Class definition!
```

Comment:
Another thing that wasn't clarified in this section, or maybe should get a little more attention, is:

What exactly does `* => *` mean?